### PR TITLE
Do not throw exceptions when we can't make a doc id

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/71460
Fixes https://github.com/dotnet/roslyn/issues/70159

### CODE REVIEW

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 1]
```diff
@@ -70,7 +77,15 @@ public static string CreateReferenceId(ISymbol symbol)
 
             if (symbol is INamespaceSymbol)
             {
-                return CreateDeclarationId(symbol);
+                // This is very odd.  For anything other than a namespace, we defer to ReferenceGenerator, which just
+                // spits out a reference.  But for a namespace, we spit out a declaration (which means we prefix with
+                // `N:`).  None of the other paths do this.  This is likely a bug, but it appears as if it has never
+                // been noticed.
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:84]: Something super odd i noticed.  This is in CreateReferenceId (which does not append the `X:` prefix on any symbol).  But for namespaces... it does.  This can't possibly be expected.  My guess is that no one actually ever tried a namespace symbol here (or tested it).

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 2]
```diff
@@ -501,7 +512,7 @@ public override bool VisitNamedType(INamedTypeSymbol symbol)
             }
         }
 
-        private class ReferenceGenerator : SymbolVisitor<bool>
+        private sealed class ReferenceGenerator : SymbolVisitor
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:515]: the bool in this type was only used to determine if any text was written when hitting a container (in order ot determine if a '.' was needed between nested types/namespaces).  We don't need that (and the bool could be confused for success/failure), so removed.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 3]
```diff
@@ -599,23 +604,20 @@ public override bool VisitArrayType(IArrayTypeSymbol symbol)
...
             }
 
-            public override bool VisitTypeParameter(ITypeParameterSymbol symbol)
+            public override void VisitTypeParameter(ITypeParameterSymbol symbol)
             {
                 if (!IsInScope(symbol))
                 {
                     // reference to type parameter not in scope, make explicit scope reference
-                    var declarer = new DeclarationGenerator(_builder);
+                    var declarer = new PrefixAndDeclarationGenerator(_builder);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:623]: this is interesting.  we will generate something like `"T:Acme.OuterType`1:`0"`, even when generating a *reference*.  It feels wonky to me.  But we have tests for this.  So i guess it's intentional?  It's the only case for references though where we emit `prefix-colon` again...

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  this was approved by API team and is ready for review.

**Comment by CyrusNajmabadi** [comment]
tagging @333fred @jjonescz

**Review by genlu — APPROVED** [review]
IDE change LGTM

File: src/Compilers/Core/CodeAnalysisTest/Symbols/DocumentationCommentIdTests.cs [hunk 4]
```diff
@@ -154,6 +156,35 @@ End Namespace
...
+                class C
+                {
+                    unsafe void M({{type}} i)
+                    {
+                        var x = i + 1;
+                    }
+                }
+                """;
+
+            var comp = CreateCSharpCompilation(text, compilationOptions: TestOptions.UnsafeDebugDll).VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/Symbols/DocumentationCommentIdTests.cs:173]: >CreateCSharpCompilation [](http://example.com/codeflow?start=23&length=23)

Can we add a "similar" test for VB?

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 5]
```diff
@@ -41,26 +41,34 @@ public new void Free(List<T> list)
...
             if (symbol == null)
             {
                 throw new ArgumentNullException(nameof(symbol));
             }
 
-            var builder = new StringBuilder();
-            var generator = new DeclarationGenerator(builder);
-            generator.Visit(symbol);
-            return builder.ToString();
+            var builder = PooledStringBuilder.GetInstance();
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:56]: >PooledStringBuilder.GetInstance() [](http://example.com/codeflow?start=26&length=33)

It looks like the switch to ```PooledStringBuilder``` is unrelated to the stated goal of the PR. We prefer to not bundle unrelated changes in a single PR. While, I don't see anything wrong in the change itself, I think it should be reverted and submitted as a separate PR.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 6]
```diff
@@ -301,68 +317,61 @@ private static string DecodePropertyName(string name, string language)
...
-        private class DeclarationGenerator : SymbolVisitor
+        /// <summary>
+        /// The <see langword="bool"/> type parameter indicates if we succeeded at writing out the declaration ID or not. 
+        /// </summary>
+        private sealed class PrefixAndDeclarationGenerator(StringBuilder builder) : SymbolVisitor<bool>
         {
-            private readonly StringBuilder _builder;
-            private readonly Generator _generator;
-
-            public DeclarationGenerator(StringBuilder builder)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:309]: >public DeclarationGenerator(StringBuilder builder) [](http://example.com/codeflow?start=12&length=50)

The switch to primary constructor is unnecessary and it makes the code harder to follow. Please revert.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 7]
```diff
@@ -301,68 +317,61 @@ private static string DecodePropertyName(string name, string language)
             return name;
         }
 
-        private class DeclarationGenerator : SymbolVisitor
+        /// <summary>
+        /// The <see langword="bool"/> type parameter indicates if we succeeded at writing out the declaration ID or not.
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:321]: >type parameter indicates [](http://example.com/codeflow?start=39&length=24)

The wording feels confusing. Type parameters do not have values. Perhaps you are referring to the result returned by the `Visit` methods.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:309]: BTW, apart from issues specifically with the primary constructor (I would ask to not use it even if that was used with a new type, probably with a very small exception), this also constitutes as a style change. Style changes are not welcome.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 8]
```diff
@@ -301,68 +317,61 @@ private static string DecodePropertyName(string name, string language)
...
             }
 
-            private class Generator : SymbolVisitor<bool>
+            private sealed class DeclarationGenerator(StringBuilder builder) : SymbolVisitor<bool>
             {
-                private readonly StringBuilder _builder;
+                private readonly StringBuilder _builder = builder;
                 private ReferenceGenerator? _referenceGenerator;
 
-                public Generator(StringBuilder builder)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:361]: >public Generator(StringBuilder builder) [](http://example.com/codeflow?start=16&length=39)

Please go back to the regular constructor

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 9]
```diff
@@ -301,68 +317,61 @@ private static string DecodePropertyName(string name, string language)
...
 
-            public override void DefaultVisit(ISymbol symbol)
-            {
-                throw new InvalidOperationException("Cannot generated a documentation comment id for symbol.");
-            }
+            /// <summary>
+            /// If we hit anything we don't know about, indicate failure.
+            /// </summary>
+            public override bool DefaultVisit(ISymbol symbol)
+                => false;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:332]: >=> false; [](http://example.com/codeflow?start=16&length=9)

This change feels fragile, there is no guarantee that every caller is going to check the return value and propagate `false`.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 10]
```diff
@@ -301,68 +317,61 @@ private static string DecodePropertyName(string name, string language)
             return name;
         }
 
-        private class DeclarationGenerator : SymbolVisitor
+        /// <summary>
+        /// The <see langword="bool"/> type parameter indicates if we succeeded at writing out the declaration ID or not. 
+        /// </summary>
+        private sealed class PrefixAndDeclarationGenerator(StringBuilder builder) : SymbolVisitor<bool>
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:323]: >PrefixAndDeclarationGenerator [](http://example.com/codeflow?start=29&length=29)

I find changes made to the visitors both fragile an unnecessary. Therefore here is what I suggest to do:
1. Revert all changes in this file, including all renames and unrelated optimizations.  
2. Rather than throwing ```InvalidOperationException```, switch to throwing a private exception type in the visitors
3. In the two/three places that use the visitors, catch the private exception and assume the result is null.
4. If you still would like to rename types and optimize StringBuilder operations, submit those as separate PRs.

**Comment by AlekseyTs** [comment]
> 

Consider placing this line next to the corresponding addition so that it is obvious that we are only changing nullability of the result.

---
Refers to: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:41 in 0021b1e. [](commit_id = 0021b1edb35d653af92f37474f00edeb1336dfa4, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 27)

File: src/Compilers/Core/CodeAnalysisTest/Symbols/DocumentationCommentIdTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/CodeAnalysisTest/Symbols/DocumentationCommentIdTests.cs:173]: Sure.  though i'm not knowledgable enough about the VB symbol model to know what would work here.  I used dynamic and pointers in c# because i knew about problems there.  But i don't think VB has either of those.  Is there another type of non-named type symbol that vb has where operators are synthesized for it?

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs]: Removed.  Will add in separate pr.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:309]: To me, this is not a style only change.  It fundamentally distinguishes to me a major difference in how an object works, amking it much easier to reason about.  In this case, i can tell that the builder is no longer a field of the generator, it is simply a parameter, and can be thought of as having captured semantics for the purposes of this helper type.  That makes reasoning about how this works much simpler, and more in line with a helper method/function/lambda.

To me, when classes are introduced for this sort of purpose, this is a much easier and simpler way to reason about things, and it (and other changes) helped reveal several problems in this area that had previously gone unnoticed.  I would prefer to keep this.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs]: tweaked docs.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:328]: there is only two external caller.  One of which is the mainline call (And checks) and the other is a caller within referencegenerator which checks as well.  It doesn't feel particularly fragile to me.  No more so than any case where a visitor (or any api) is used and has a return value.  I can doc the type even more heavily to indicate that it might fail if that helps.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:325]: I would prefer to not revert the renames.  They were very helpful for actually determining what the type was doing and seeing where things were odd.  Keepign clear names (including the distinction on if the DocID prefix is included or not very much helped with clarity and diagnosability here.

2. I would prefer to not use exceptions in normal non-exceptional codepaths.  They are expensive under a debugger, and often end up interfering with teh debugging experience (breaking during normal operations).  We get reports from VS and other teams about pain points here when exceptions are thrown in codepaths that done need it.

3. There seems to be only one real codepath that flows null upwards.  And it was clearly identified by either being DefaultVisit, or a recursive call to Visit.  This was easy to audit and write code for.  So i'd prefer to just use flowing that boolean upwards versus needing to use an exception with all the machinery that entails.

4. I've pulled out the stringbuilder change.  I can do that in anotehr PR.  HOwever, the renames and other structural changes were very relevant in the course of auditing this type.  IMO, had those changes been there originally, i think it would have been less likely that these issues would have arisen. As such, i do not view them as stylistic, but as inherent to good design and impl for this space.

**Comment by CyrusNajmabadi** [comment]
> Consider placing this line next to the corresponding addition so that it is obvious that we are only changing nullability of the result.

SGTM.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:79]: Note: i'm fine changing this.  But it was out of scope for the problem i was trying to address, so i left it alone.  However, it very much was found to be a problem thanks to the approach to auditing taken in this pr.

**Comment by CyrusNajmabadi** [comment]
Improvements pulled out into https://github.com/dotnet/roslyn/pull/71533

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 10 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:325]: Sorry, but my position didn't change. We can discuss offline in more details.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:309]: Moved to cleanup pr.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:361]: Moved to cleanup pr.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 11]
```diff
@@ -467,41 +524,19 @@ private void AppendParameters(ImmutableArray<IParameterSymbol> parameters)
...
-                        return false;
-                    }
-
-                    if (this.Visit(symbol.ContainingSymbol))
-                    {
-                        _builder.Append(".");
-                    }
-
-                    _builder.Append(EncodeName(symbol.Name));
+                    AppendNamespace(symbol, _builder);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:527]: move to helper that cannot fail.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 12]
```diff
@@ -467,41 +524,19 @@ private void AppendParameters(ImmutableArray<IParameterSymbol> parameters)
...
-
-                    _builder.Append(EncodeName(symbol.Name));
-
-                    if (symbol.TypeParameters.Length > 0)
-                    {
-                        _builder.Append("`");
-                        _builder.Append(symbol.TypeParameters.Length);
-                    }
-
+                    AppendNamedType(symbol, _builder);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:533]: move to helper that cannot fail.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 13]
```diff
@@ -301,64 +312,113 @@ private static string DecodePropertyName(string name, string language)
...
-                _generator.Visit(symbol);
+                return _generator.Visit(symbol);
             }
 
-            public override void VisitNamespace(INamespaceSymbol symbol)
+            public override bool VisitNamespace(INamespaceSymbol symbol)
             {
                 _builder.Append("N:");
-                _generator.Visit(symbol);
+                AppendNamespace(symbol, _builder);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:369]: move to helper that cannot fail.  as such, we can guarantee that this always succeds in the return statement (and doc this in the constracts).

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 14]
```diff
@@ -301,64 +312,113 @@ private static string DecodePropertyName(string name, string language)
...
+                return true;
             }
 
-            public override void VisitNamedType(INamedTypeSymbol symbol)
+            public override bool VisitNamedType(INamedTypeSymbol symbol)
             {
                 _builder.Append("T:");
-                _generator.Visit(symbol);
+                _generator.VisitNamedType(symbol);
+                return true;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:377]: move to helper that cannot fail.  as such, we can guarantee that this always succeds in the return statement (and doc this in the constracts).

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 15]
```diff
@@ -301,64 +312,113 @@ private static string DecodePropertyName(string name, string language)
...
+            }
+
+            private static void AppendNamespace(INamespaceSymbol symbol, StringBuilder builder)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:380]: moved the visit methods into these explicit namespace/named-type helpers.  it makes it much clearer that there are no failure paths here.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 16]
```diff
@@ -373,63 +433,60 @@ private ReferenceGenerator GetReferenceGenerator(ISymbol typeParameterContext)
...
                 }
 
+                /// <summary>
+                /// If we hit anything we don't know about, indicate failure.
+                /// </summary>
                 public override bool DefaultVisit(ISymbol symbol)
-                {
-                    throw new InvalidOperationException("Cannot generated a documentation comment id for symbol.");
-                }
+                    => false;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:440]: Crux of the change.  no throwing exceptions for expected cases.  instead, we just indicate we cannot proceed.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs]: Moved to a non-flow based approach to this determination.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs]: Other auditing parts have been pulled out to a separate pr.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler this is ready for review.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 17]
```diff
@@ -301,22 +314,39 @@ private static string DecodePropertyName(string name, string language)
             return name;
         }
 
-        private class DeclarationGenerator : SymbolVisitor
+        /// <summary>
+        /// The <see langword="bool"/> generic type argument indicates if we succeeded at writing out the declaration ID
+        /// or not.  Callers should only call into <see cref="SymbolVisitor{TResult}.Visit(ISymbol?)"/> and should check
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:319]: It looks like this statement is product of earlier iterations and no longer relevant.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:319]: indeed!  will fix.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 18]
```diff
@@ -301,22 +314,39 @@ private static string DecodePropertyName(string name, string language)
...
+        /// <see cref="Failed"/> to see if it failed (in the case of an arbitrary symbol) or that it produced an
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:320]: ><see cref="Failed"/>  [](http://example.com/codeflow?start=12&length=21)

It looks like the flag is not going to be reset for a subsequent ```Visit``` calls. It might be worth mentioning the fact in this comment,

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 19]
```diff
@@ -344,21 +374,59 @@ public override void VisitMethod(IMethodSymbol symbol)
             public override void VisitNamespace(INamespaceSymbol symbol)
             {
                 _builder.Append("N:");
-                _generator.Visit(symbol);
+                AppendNamespace(symbol, _builder);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:377]: >AppendNamespace(symbol, _builder); [](http://example.com/codeflow?start=16&length=34)

This change feels outside of scope of this PR and, I think,  should be excluded. I suggest opening a dedicated issue, and, optionally, creating a separate PR with appropriate tests added (those that fail without the change).

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 20]
```diff
@@ -344,21 +374,59 @@ public override void VisitMethod(IMethodSymbol symbol)
...
             }
 
             public override void VisitNamedType(INamedTypeSymbol symbol)
             {
                 _builder.Append("T:");
-                _generator.Visit(symbol);
+                AppendNamedType(symbol, _builder);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:383]: >AppendNamedType(symbol, _builder); [](http://example.com/codeflow?start=16&length=34)

Similar comment as for the previous method, bath can be part of the same issue.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 18 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs]: sure!

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 21]
```diff
@@ -344,21 +374,59 @@ public override void VisitMethod(IMethodSymbol symbol)
...
+                builder.Append(EncodeName(symbol.Name));
+
+                if (symbol.TypeParameters.Length > 0)
+                {
+                    builder.Append('`');
+                    builder.Append(symbol.TypeParameters.Length);
+                }
+            }
+
+            private sealed class DeclarationGenerator : SymbolVisitor
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:387]: > private sealed class DeclarationGenerator : SymbolVisitor [](http://example.com/codeflow?start=11&length=58)

In my opinion the changes from this point to the end of the file that are not related to the introduction of the `Failed` flag in this and the enclosing type are unrelated to the stated goal of the PR. I think they should be reverted. Perhaps with exception of the type rename.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 20 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs]: wfm.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 21 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs]: wfm.

File: src/Compilers/Core/Portable/DocumentationCommentId.cs [hunk 22]
```diff
@@ -615,7 +652,7 @@ public override bool VisitTypeParameter(ITypeParameterSymbol symbol)
                 if (!IsInScope(symbol))
                 {
                     // reference to type parameter not in scope, make explicit scope reference
-                    var declarer = new DeclarationGenerator(_builder);
+                    var declarer = new PrefixAndDeclarationGenerator(_builder);
                     declarer.Visit(symbol.ContainingSymbol);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DocumentationCommentId.cs:656]: >declarer.Visit(symbol.ContainingSymbol); [](http://example.com/codeflow?start=20&length=40)

We don't expect this call to fail? Consider asserting the fact.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 56).

File: src/Compilers/Core/Portable/DocumentationCommentId.cs (hunk 22 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/DocumentationCommentId.cs:656]: SGTM.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 57). Thanks!

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler for another pair of eyes.  Tnx!

**Review by 333fred — APPROVED** [review]
